### PR TITLE
Add CI/CD workflows around poetry

### PR DESCRIPTION
**Pull request by djperrefort**
I've implemented a few CI/CD workflows for testing and documentation. The implementation follows an architecture I designed for the CRC a few months ago. Overall we've had great success using a very similar configuration.

The sections below outline what pieces of the automation run with each trigger. I made a few changes to help facilitate the automation:
1.  As discussed earlier today, I updated the packaging to use a `pyproject.toml` file. Note the version number is set at `0.0.0`. This number will now be set dynamically at the time of release (more details below). Where necessary, the package fetches the value directly from the installed metadata instead of a `VERSION.txt` file. 
2. I added a placeholder test suite. Nothing is actually being tested, but the CI needs something to run.
3. I made some edits to the docs to resolve build warnings that were causing the build CI to fail. To be clear, the CI is designed to fail on these types of warnings.
4. I had some issues getting things working with Python > 3.7 - mostly with dependency resolution/conflicts. I updated the version specification on the dependencies to use recent versions, switched to 3.8+, and the problem went away. If there are any additional bugs/issues, we will catch them while filling in the test suite. Requiring 3.7 isn't sustainable in the long run anyways.

### On Every Push

Pushes happen frequently, so the automation on this trigger is minimal. 
On this trigger we focus on "testing often and failing fast".

- The test suite is executed on Ubuntu against Python 3.8 through 3.11.
- Code coverage results are reported to the Codacy QA platform

Note: Additional test environments (e.g., multiple operating systems) can be easily added as needed or desired.

### On Every PR

Workflows triggered by PR are specifically designed to be useful as quality gates.
Once implemented, passing individual workflows can be set as a requirements before merging into `main`.  

- The test suite is executed
- The Codacy QA platform checks for any quality issues
- The documentation is compiled to check for build errors

### On Every Release

Releases rerun all testing workflows.
This catches implicit bugs that can work their way into `main` via various merging patterns.

- The test suite is executed
- The documentation is compiled to check for build errors

If all the above workflows pass, the package/documentation is published.
The version number is pulled directly from the release name.
For example, release `v0.3.1` tags the package as version `0.3.1`.

- The documentation is published to GitHub pages
- The package is published to PyPI.org

Here is an example of a nearly identical pipeline running in the real world: https://github.com/pitt-crc/quota_notifier/actions/runs/4255035264

### Manual Triggers

The test suite, documentation build, and documentation publication can all be triggered directly at any time.
Publishing a new version cannot - a release is required to ensure quality (i.e. to ensure the full CD pipeline runs).

### Cron

Once a month the GitHub `dependabot` utility will check the version of each CI step we are running and submit a PR updating them to the latest version. This ensures the CI tool chain stays up to date.

### CODE REVIEW

**Review by troyraen — APPROVED** [review]
LGTM, thanks.

#6 makes major updates to the docs. Would you mind checking them and let me know if there are any issues they will cause when merged with this?

File: pyproject.toml [hunk 1]
```diff
@@ -1,6 +1,46 @@
...
+matplotlib = "^3.5"
+numpy = "^1.20.0"
+pandas = "^1.3"
+tabulate = "^0.9"
+
+[tool.poetry.group.docs]
+optional = true
+
+[tool.poetry.group.docs.dependencies]
+docutils = "<0.17" # render lists properly
```
[the remark this answers is not in the document]
Comment (wmwv) [reply pyproject.toml:37]: I appreciate that the `docutils = "<0.17"` was already in the previous `requirements.txt` that is being replaced.

[the remark this answers is not in the document]
Comment (djperrefort) [reply pyproject.toml:37]: No Idea. I was just being consistent with `requirements.txt`. I think it is worth an editing pass after the PR to bring some of the dependencies up to date (I volunteer).

**Comment by djperrefort** [comment]
@troyraen I didn't formally test #6 but it looks OK. It will need to be brought up to date after the PR goes through, but I'm not expecting many conflicts.

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (wmwv) [reply pyproject.toml:37]: Sounds good.  Thanks.

**Pull request merged**